File: elixir/rna-transcription/rna_transcription.exs
defmodule RNATranscription do
  @doc """
  Transcribes a character list representing DNA nucleotides to RNA

  ## Examples

  iex> RNATranscription.to_rna('ACTG')
  'UGAC'
  """
  @spec to_rna([char]) :: [char]
  def to_rna(dna) do
    for nucleotide <- dna, do: transcribe(nucleotide)
  end

  defp transcribe(?A), do: ?U
  defp transcribe(?C), do: ?G
  defp transcribe(?T), do: ?A
  defp transcribe(?G), do: ?C
end

File: erlang/rna-transcription/src/rna_transcription.erl
-module(rna_transcription).

-export([to_rna/1, test_version/0]).

-type nucleotides() :: [$A | $T | $C | $G].

%% API

-spec to_rna(nucleotides()) -> nucleotides().
to_rna(Strand) ->
    [transcribe(N) || N <- Strand].

-spec test_version() -> integer().
test_version() ->
    3.

%% Internal

transcribe($G) -> $C;
transcribe($C) -> $G;
transcribe($T) -> $A;
transcribe($A) -> $U.
